Source/NSDictionaryController.m

```diff
@@ -0,0 +1,480 @@
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import <Foundation/NSArray.h>
+#import <Foundation/NSDictionary.h>
+#import <Foundation/NSException.h>
+#import <Foundation/NSIndexSet.h>
+#import <Foundation/NSKeyValueObserving.h>
+#import <Foundation/NSPropertyList.h>
+#import <Foundation/NSString.h>
+
+#import "AppKit/NSDictionaryController.h"
+#import "AppKit/NSKeyValueBinding.h"
+
+#import "GSBindingHelpers.h"
+#import "GSFastEnumeration.h"
+
+// Private methods for NSString and NSDictionary...
+@interface NSDictionary (GSPrivate_DictionaryToStrings)
+
+- (NSString *) _stringsFromDictionary;
+
+@end
+
+@implementation NSDictionary (GSPrivate_DictionaryToStrings)
+
+- (NSString *) _stringsFromDictionary
+{
+  NSEnumerator *en = [self keyEnumerator];
+  NSString *k = nil;
+  NSString *result = @"";
+
+  while ((k = [en nextObject]) != nil)
+    {
+      NSString *v = [self objectForKey: k];
+      result = [result stringByAppendingString:
+			 [NSString stringWithFormat: @"\"%@\" = \"%@\";\n", k, v]];
+    }
+
+  return result;
+}
+
+@end
+
+@interface NSString (GSPrivate_StringsToDictionary)
+
+- (NSDictionary *) _dictionaryFromStrings;
+
+@end
+
+@implementation NSString (GSPrivate_StringsToDictionary)
+
+- (NSDictionary *) _dictionaryFromStrings
+{
+  NSError *error = nil;
+  NSDictionary *dictionary =
+    [NSPropertyListSerialization
+      propertyListWithData: [self dataUsingEncoding: NSUTF8StringEncoding]
+		   options: NSPropertyListImmutable
+		    format: NULL
+		     error: &error];
+  if (error != nil)
+    {
+      NSLog(@"Error reading strings file: %@", error);
+    }
+
+  return dictionary;
+}
+
+@end
+
+@interface NSDictionaryController (GSPrivate_BuildArray)
+
+- (NSArray *) _buildArray: (NSDictionary *)content;
+
+@end
+
+@implementation NSDictionaryController (GSPrivate_BuildArray)
+
+- (NSArray *) _buildArray: (NSDictionary *)content
+{
+  NSArray *allKeys = [content keysSortedByValueUsingSelector: @selector(compare:)];
+  NSMutableArray *result = [NSMutableArray arrayWithCapacity: [allKeys count]];
+
+  FOR_IN(id, k, allKeys)
+    {
+      NSDictionaryControllerKeyValuePair *kvp =
+	AUTORELEASE([[NSDictionaryControllerKeyValuePair alloc] init]);
+      id v = [content objectForKey: k];
+      NSString *localizedKey = [_localizedKeyDictionary objectForKey: k];
+
+      [kvp setLocalizedKey: localizedKey];
+      [kvp setKey: k];
+      [kvp setValue: v];
+      [kvp setExplicitlyIncluded: NO];
+
+      if ([_excludedKeys containsObject: k])
```

Comment:
This could be checked first before creating and setting kvp.

Reply:
Yes, done.